### ISSUE TITLE
Equality Predicates

### ISSUE THREAD

**Issue by sjl** [danshapero/cl-dict#2]
It would be nice to have `hash-set=` and `hash-dict=` functions.  I can do `(zerop (set-size (set-symmetric-diff a b)))` but that seems awkward...

**Comment by danshapero** [comment]
Good idea! I can get to this in a few weeks' time, but if you implement it yourself before then I'd be happy to accept the PR.

**Comment by sjl** [comment]
I might have time, might not.

If I do, what should happen when comparing things with different `:test`s?  Should it just be an error, or just return `nil` to mean that they're not equal?  I'd probably lean toward an error personally...

**Comment by danshapero** [comment]
That sounds like an error to me too. Do you know how to add conditions, restarts, that sort of thing? I haven't used them before but it would be a good excuse for me to learn.

**Comment by danshapero** [comment]
@sjl There's also the question of how you compare equality for the values in the dictionary. I can imagine some use case where one day you want to compare two dictionaries with string values using `equal`, which is case-sensitive, and the next day you want to use `equalp` which is not. In that case, it would make sense to pass the value equality predicate as an argument:

``` common-lisp
(defun hash-dict= (dict1 dict2 &optional (value-eq-p 'equalp))
   ;; magic goes here
  )
```

**Comment by sjl** [comment]
Hm, yeah.  Maybe call it `value-test` to match the vanilla `test` for the keys?

**Comment by danshapero** [comment]
@sjl I've implemented the equality predicates in [this branch](https://github.com/danshapero/cl-hamt/tree/eq), thoughts or suggestions?

**Comment by sjl** [comment]
Just skimmed over it a bit.  I'm not super familiar with the inner workings of HAMTs but it seems pretty sane.

You might want to consider normalizing the `test` designator upon construction now that you're comparing them with `eq`.  Right now I think `(set-eq (empty-set :test 'equal) (empty-set :test #'equal))` will fail.  As a bonus, doing the function lookup once at construction time if the designator is a symbol would save a `(symbol-function ...)` lookup at every `funcall` of it.

Then again, resolving the function at construction time means users can't redefine the equality function and expect it to take effect for existing HAMTs... but I feel like it's probably a Bad Idea™ to redefine your HAMTs' equality predicate on the fly anyway.

**Comment by danshapero** [comment]
Wow good catch! I made a fix in 1b521213a3b0b20a4 , is that what you had in mind?

**Comment by sjl** [comment]
Yeah something like that.  If you want a slightly friendlier error message when some terrible user passes something that's just totally bonkers as a `:test` you could do:

```lisp
(ctypecase test
  (function test)
  (symbol (symbol-function test))
```

And then passing something stupid like `"equal"` would give a nicer error:

```
  "equal" fell through CTYPECASE expression. Wanted one of (SYMBOL FUNCTION).

restarts (invokable by number or by possibly-abbreviated name):
  0: [STORE-VALUE] Supply a new value for TEST.
  1: [ABORT      ] Exit debugger, returning to top level.
```

But I wouldn't blame you if you didn't bother -- that's pretty unlikely in practice (I hope).

**Comment by danshapero** [comment]
Awesome, I used your way. Thanks for the suggestions!

Closed by fc9e292bc53.
